Spec the pattern string parser.

wanderview opened the pull request:
<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/pull/57.html" title="Last updated on Jul 29, 2021, 6:27 PM UTC (7a58f60)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/57/8cb3403...7a58f60.html" title="Last updated on Jul 29, 2021, 6:27 PM UTC (7a58f60)">Diff</a>

wanderview commented:
@domenic PTAL.  The c++ source code for this PR can be found here:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/liburlpattern/parse.cc

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
 
+<h4 id=parts>Parts</h4>
+
+A <dfn>part list</dfn> is a [=list=] of zero or more [=part=] [=structs=].
```
```suggestion
A <dfn>part list</dfn> is a [=list=] of zero or more [=parts=].
```

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+<h4 id=options-header>Options</h4>
+
+An <dfn>options</dfn> [=struct=] contains different settings that control how [=/pattern string=] behaves.  These options originally come from [path-to-regexp](https://github.com/pillarjs/path-to-regexp).  Not all are utilized when constructing a {{URLPattern}}, but they are maintained here to make it easier to stay aligned with [path-to-regexp](https://github.com/pillarjs/path-to-regexp).
```
Sorry, but you should not include in the spec behavior that is not observable.

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">encoding callback</dfn>, a [=/encoding callback=], initially null.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">options</dfn>, a [=/options=], initially null.
```
Optional, but perhaps a bit clearer: if there are mandatory fields that are always set upon creation, you can say "which must be set upon creation". Then when creating you can say something like "a new pattern parser whose encoding callback is X and options is Y".

This makes it obvious which of these defaults are real defaults that are going to stick around. Right now it seems like some of these defaults are arbitrary and could be anything because you always overwrite them.

domenic replied:
Same goes for the part struct it seems.

wanderview replied:
The option defaults actually come from path-to-regexp, but I can drop them since we don't use them.

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+The <dfn>full wildcard regexp value</dfn> is the string "`.*`".
+
+<div algorithm>
```
It's a style thing, but I prefer putting the most important algorithms (in this case "parse a pattern") at the top of the section, and putting auxiliary helpers afterward.

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+<div algorithm>
+To <dfn>try to consume a token</dfn> given a [=pattern parser=] |parser| and [=token/type=] |type|:
+
+  1. Let |next token| be |parser|'s [=pattern parser/token list=][|parser|'s [=pattern parser/index=]].
```
It's not clear from what I've read so far that the index will always point to a token that exists. Indeed I guess it's not true in the initial state (empty list/index 0).

It might be good to either assert that, if it's true for all callers, or add a note after the definition of index saying that that's always true once some initial setup is performed.

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+  1. Let |result| be the empty string.
+  1. Let |done| be false.
+  1. While |done| is false:
```
Can you avoid the done variable by just using `[=iteration/break=]` and "While true:"?

wanderview replied:
I really want a `do {} while()` loop.  What is the best spec construct for that?

domenic replied:
I don't understand why a do/while would be necessary here. If the first token ends up being null, you do nothing, right? So just a normal while loop would work.

wanderview replied:
I don't think that's the case?

https://source.chromium.org/chromium/chromium/src/+/main:third_party/liburlpattern/parse.cc;l=79-87;drc=ccecd870311e5326a355715201778816c841e8f0

I guess I can just duplicate the two "try to consume" steps to get the token before the loop.  Just using `While true:` is probably easier, though.

domenic replied:
I'm still confused. Here is my proposal:

While true:

1. Let token be the result of running try to consume a token given parser and "char".
1. If token is null, then set token to the result of running try to consume a token given parser and "escaped-char".
1. If token is null, then break.
1. Append token's value to the end of result.

Let's say both "char" and "escaped-char" will return null for the first token. Then:

- in your version we run steps 1, 2 which have side effects. 3 sets done to true. 4 does nothing since it's "else". Then we go back to the top of the loop where done is false, so we're done.

- in my version we run steps 1, 2. 3 breaks. So we're done.

domenic replied:
(I guess it's not clear whether there are actually side effects... it seems like the index doesn't advance in that case.)

wanderview replied:
Yea, your proposal works.  Its equivalent to the "do while" without the break statement.

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+    1. Else if |modifier token|'s [=token/value=] is "`+`" then set |modifier| to "<a for=part/modifier>`one-or-more`</a>".
+  1. If |name token| is null and |regexp or wildcard token| is null and |modifier| is "<a for=part/modifier>`none`</a>":
+    1. Append |prefix| to the end of |parser|'s [=pattern parser/pending fixed value=].
+    1. Return
```
```suggestion
    1. Return.
```

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+  1. Run [=maybe add a part from the pending fixed value=] given |parser|.
+  1. Let |part| be a new [=part=].
+  1. If |name token| is null and |regexp or wildcard token| is null:
+    1. [=Assert=] |suffix| is the empty string.
```
```suggestion
    1. [=Assert=]: |suffix| is the empty string.
```

domenic commented on spec.bs:
```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+    1. Let |name token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`name`</a>".
+    1. Let |regexp or wildcard token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`regexp`</a>".
+    1. If |name token| is null and |regexp or wildcard token| is null:
+      1. Set |regexp or wildcard token| to the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`asterisk`</a>".
```
Probably this doesn't need to be nested.

wanderview replied:
Done.  And I factored it out to a separate algorithm since we need to do these steps at least twice.

wanderview commented:
@domenic PTAL.  I addressed your action items.  I also added a bit about the `segment wildcard regexp` that I forgot before.  And I tried adding some notes based on the c++ comments.  Maybe some of them are not helpful enough.  Let me know what you think.  Thanks.

domenic commented on spec.bs:
```diff
@@ -719,8 +660,341 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">encoding callback</dfn>, a [=/encoding callback=], that must be set upon creation.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">options</dfn>, a [=/options=], that must be set upon creation.
```
It looks like this field is never referenced. Will it be? Or maybe it's just necessary in C++ where closures are a pain, but in the spec you didn't need it.

wanderview commented on spec.bs:
```diff
@@ -719,8 +660,341 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+      1. Run [=add a part=] given |parser|, |prefix|, |name token|, |regexp or wildcard token|, the empty string, and |modifier token|.
+      1. [=Continue=].
+    1. Let |fixed token| be |char token|.
+      <p class=note>If there was no matching group, then we need to buffer an fixed text.  We want to collect as much text as possible before adding it as a "<a for=part/type>`fixed-text`</a>" [=part=].
```
"any fixed text"

wanderview commented on spec.bs:
```diff
@@ -719,8 +660,341 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+An [=/options=] has an associated <dfn for=options>delimiter list</dfn>, a string, which must be set upon creation.  Each code point in the [=options/delimiter list=] is treated as a segment separator.  This is used, for example, for determining how far a `:foo` named group should match by default.
+
+An [=/options=] has an associated <dfn for=options>prefix list</dfn>, a string, which must be set upon creation.  Each code point in the [=options/prefix list=] is treated as an automatic prefix if found immediately preceding a match group.  This matters when a match group is modified to be optional or repeating.  For example, if [=options/prefix list=] is "`/`" then "`/foo/:bar?/baz`" will treat ...
```
We never use it with more than one code point, so I changed it to `prefix code point`.

wanderview commented on spec.bs:
```diff
@@ -719,8 +660,341 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+An <dfn>options</dfn> [=struct=] contains different settings that control how [=/pattern string=] behaves.  These options originally come from [path-to-regexp](https://github.com/pillarjs/path-to-regexp).  We only include the options that are modified within the URLPattern specification and exclude the other options.  For the purposes of comparison, this specification acts like [path-to-regexp](h ...
+
+An [=/options=] has an associated <dfn for=options>delimiter list</dfn>, a string, which must be set upon creation.  Each code point in the [=options/delimiter list=] is treated as a segment separator.  This is used, for example, for determining how far a `:foo` named group should match by default.
```
Rewrote this to be `delimiter code point` and added an example.

wanderview commented:
Updated.  @domenic PTAL.

domenic reviewed (approved):
Looks great; the examples and string vs. code point change is very helpful.
